src/Equations/SPIHandler.cpp

```diff
@@ -324,6 +553,21 @@ void SPIHandler::Rebuild(real_t dt){
     if(spi_deposition_mode==OptionConstants::EQTERM_SPI_DEPOSITION_MODE_LOCAL){
         CalculateTimeAveragedDeltaSourceLocal(depositionProfilesAllShards);
         
+        // Calculate drift (if any)
+        if(spi_shift_mode==OptionConstants::EQTERM_SPI_SHIFT_MODE_ANALYTICAL){
+            YpConversion();
+            for(len_t ip=0;ip<nShard;ip++){
+                if (YpPrevious[ip]>0 && irp[ip]<nr){
+                    AssignShardSpecificParameters(ip);
+                    AssignComputationParameters(ip);
+                    AssignTimeParameters(ip);
+                    shift_r[ip] = Deltar(ip);
+                    if(shift_r[ip] + xp[ip]>nr*Dr)
+                        shift_r[ip] = nr*Dr-xp[ip];
+                    nbrShiftGridCell[ip]=std::round(shift_r[ip]/Dr);
+                }
+            }
+        }
         // Shift the deposition profile
         for(len_t ip=0;ip<nShard;ip++){
             if(rCoordPNext[ip]>rCoordPPrevious[ip]){
```

Comment:
I know this is mostly on me, but the loop condition here should be ir<nr-nbrShiftGridCell[ip] (before we have only used shifts of one grid cell and didn't plan well enough for the future...). Also, one should set depositionProfilesAllShards[ir*nShard+ip] = 0 for ir>nbrShiftGridCell[ip] (and make the corresponding changes for the rCoordPNext[ip]<rCoordPPrevious[ip]-case).